File: api/project/create-project.js
const Project = require("../../models/project");
const buildRD = require("../../utils/build-response-data");
const { checkMethod } = require("../../utils");
const ObjectId = require("mongoose").Types.ObjectId;

module.exports = async function createProject(request, response) {
    if (checkMethod(request, response, "POST")) return;

    const { userId, name, description } = request.body;
    if (!userId) {
        response.status(200).json(buildRD.error("User id is required"));
        return;
    }
    if (!name) {
        response.status(200).json(buildRD.error("Name is required"));
        return;
    }

    try {
        const project = await Project.findOne({
            name,
            owner: new ObjectId(userId),
        });
        if (project) {
            response.status(200).json(buildRD.error("Project already exists"));
            return;
        }
        const createRes = await Project.create({
            name,
            description,
            owner: new ObjectId(userId),
        });
        if (!createRes) {
            response
                .status(200)
                .json(buildRD.error("Failed to create project"));
            return;
        }
        response.status(200).json(
            buildRD.success({
                id: createRes._id,
                name: createRes.name,
                description: createRes.description,
                owner: createRes.owner,
                createdAt: createRes.created_at,
                updatedAt: createRes.updated_at,
            })
        );
    } catch (error) {
        response.status(200).json(buildRD.error(error.message));
        return;
    }
};

File: apis/project/createProject.go
package apis

import (
	"naotodoserver/apis"
	"naotodoserver/core"
	"naotodoserver/models"

	"github.com/gin-gonic/gin"
)

type CreateProjectHandlerV1DTO struct {
	Name        string `json:"name" binding:"required"`
	Description string `json:"description"`
}

func CreateProjectHandlerV1(ctx *gin.Context) {
	// 定义 DTO
	var dto CreateProjectHandlerV1DTO

	// 获取用户 ID
	userId, exist := ctx.Get("userId")
	if !exist {
		apis.Failure(ctx, apis.ResponseData{
			Code:    20011,
			Message: "用户凭证无效",
			Data:    nil,
		})
		return
	}

	// 获取属性
	if err := ctx.ShouldBind(&dto); err != nil {
		apis.Failure(ctx, apis.ResponseData{
			Code:    20012,
			Message: "参数错误",
			Data:    nil,
		})
		return
	}

	// 检查属性值
	if dto.Name == "" {
		apis.Failure(ctx, apis.ResponseData{
			Code:    20013,
			Message: "清单名称不能为空",
			Data:    nil,
		})
		return
	}

	// 创建记录
	var projectPreference = &models.ProjectPreference{
		UserId:     userId.(int64),
		ViewType:   "table",
		GetOptions: "{\"limit\": 20}",
		Columns:    "{\"createdAt\":false,\"updatedAt\":false,\"description\":true,\"state\":true,\"priority\":true,\"tags\":true,\"startAt\":false,\"endAt\":true}",
	}
	var projectRaw = &models.Project{
		Name:        dto.Name,
		Description: dto.Description,
		UserId:      userId.(int64),
		Preference:  projectPreference,
	}
	result := core.DB.Create(&projectRaw)
	if result.Error != nil || projectRaw.ID == 0 {
		apis.Failure(ctx, apis.ResponseData{
			Code:    20014,
			Message: "创建清单失败",
			Data:    nil,
		})
		return
	}

	// 返回成功结果
	apis.Success(ctx, apis.ResponseData{
		Code:    20010,
		Message: "创建清单成功",
		Data:    ToProjectResponse(projectRaw),
	})
}
